Added Shopping Cart to online book project.

ponomvrenko opened the pull request:
It enables to add a specific book to shopping cart.

ponomvrenko committed:
Created CartItem & ShoppingCart models.

ponomvrenko committed:
CartItem & ShoppingCart:
Improved table naming, added JoinColumn's for entity relations + naming inside DB. Also added soft delete & isDeleted fields.

ponomvrenko committed:
CartItem & ShoppingCart:
- Added nullable = false constraints.

ponomvrenko committed:
liquibase:
- Added shopping_carts & cart_items tables.

ponomvrenko committed:
Added new HashSet<> to cartItems Set.

ponomvrenko committed:
Added dto's.

ponomvrenko committed:
Added repositories.

ponomvrenko committed:
Added mappers.

ponomvrenko committed:
Added endpoints to ShoppingCartController and created ShoppingCartService layer.

ponomvrenko committed:
Checkstyle fixes in ShoppingCartServiceImpl

ponomvrenko committed:
EntityGraph was added to improve the GET endpoint

OlexVoit commented on src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Override
+    @Transactional
+    public ShoppingCartResponseDto addToShoppingCart(
+            CartItemRequestDto requestDto,
+            Long userId
+    ) {
+        CartItem cartItem = new CartItem();
+        cartItem.setQuantity(requestDto.getQuantity());
+
+        Long bookId = requestDto.getBookId();
+        Book book = bookRepository.findById(bookId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find book by this ID: " + bookId
+                ));
+        cartItem.setBook(book);
+
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find user by this ID: " + userId
+                ));
+        ShoppingCart shoppingCartFromDB =
+                shoppingCartRepository.findByUserId(userId)
+                        .orElseGet(() -> {
+                            ShoppingCart shoppingCart = new ShoppingCart();
+                            shoppingCart.setUser(user);
+                            shoppingCartRepository.save(shoppingCart);
+                            return shoppingCart;
+                        });
+        cartItem.setShoppingCart(shoppingCartFromDB);
+        cartItemRepository.save(cartItem);
+        shoppingCartFromDB.getCartItems().add(cartItem);
+
+        return shoppingCartMapper.toDto(shoppingCartFromDB);
```
very long. you have to split the method

OlexVoit commented on src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+                            return shoppingCart;
+                        });
+        cartItem.setShoppingCart(shoppingCartFromDB);
+        cartItemRepository.save(cartItem);
```
we cannot address the cartItemRepository from the shoppingCartService

ponomvrenko replied:
Hi, but at the lecture we talked about these methods and as I understand it's ok to use cartItem repository inside shoppingCartService.

OlexVoit commented on src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+        ShoppingCart shoppingCart = findShoppingCartByUserId(userId);
+        CartItem cartItem =
+                findCartItemByShoppingCartIdAndCartItemId(cartItemId, shoppingCart);
+        cartItemRepository.delete(cartItem);
```
i assume that you need to create a CartItemService and implement the logic associated with the CartItem

ponomvrenko replied:
Including all the shopping cart logic in one place was my priority, so I did not want to create a separate service for cart items.

ponomvrenko replied:
My beliefs are based on the task's conditions and Sofia's lectures.

OlexVoit commented on src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+        cartItemRepository.delete(cartItem);
+    }
+
+    private CartItem findCartItemByShoppingCartIdAndCartItemId(
```
the same

OlexVoit reviewed (changes requested):
GJ)
You need to create CartItemService and implement logic

ponomvrenko committed:
- Added admin role to ShoppingCartController
- Fix findShoppingCartByUserId by removing redundant shoppingCartId parameter.
- Optimized addToShoppingCart method.

ponomvrenko committed:
Removed admin role from endpoints in ShoppingCartController

ponomvrenko committed:
Optimized roles inside endpoints in ShoppingCartController

ponomvrenko committed:
Fixed naming of method in ShoppingCartRepository.

d1sam commented on src/main/java/com/example/onlinebookstore/controller/ShoppingCartController.java:
remove unnecessary prefix 'ROLE_' from hasRole

ponomvrenko replied:
Done!

d1sam commented on src/main/java/com/example/onlinebookstore/model/CartItem.java:
```diff
@@ -0,0 +1,40 @@
+    private ShoppingCart shoppingCart;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

d1sam commented on src/main/java/com/example/onlinebookstore/model/CartItem.java:
```diff
@@ -0,0 +1,40 @@
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

vovasalyha replied:
You marked this as resolved but didn't fix it. Why?

ponomvrenko replied:
@vovasalyha my fault, I just saw that Dima gave me two identical comments, before that I thought he was only one

d1sam commented on src/main/java/com/example/onlinebookstore/repository/CartItemRepository.java:
```diff
@@ -0,0 +1,7 @@
+import com.example.onlinebookstore.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
```
```suggestion
@Repository
public interface CartItemRepository extends JpaRepository<CartItem, Long> {
```

d1sam commented on src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,106 @@
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find user by this ID: " + userId
+                ));
+        ShoppingCart shoppingCartFromDB = shoppingCartRepository.findByUserIdWithCartItems(userId)
```
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserIdWithCartItems(userId)
```

d1sam commented on src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,106 @@
+                ));
+    }
+
+    private CartItem cartItemFormationWithoutShoppingCart(CartItemRequestDto requestDto) {
```
```suggestion
    private CartItem createCartItemWithoutShoppingCart(CartItemRequestDto requestDto) {
```

d1sam reviewed (changes requested):
good job, but there are some change should be done!

ponomvrenko committed:
Removed 'ROLE_' from `@PreAuthorize` in endpoints. Also fixes naming of some methods and variables.

d1sam commented on src/main/java/com/example/onlinebookstore/controller/BookController.java:
it's not related to this hw.

d1sam commented on src/main/java/com/example/onlinebookstore/controller/CategoryController.java:
it's not related to this hw.

d1sam reviewed (changes requested):
please remove changes not related to this hw.

ponomvrenko committed:
Returned all 'ROLE_' to endpoints that are not relevant to this PR

vovasalyha reviewed (approved):
Approving, but if you are marking some comment as resolved without making changes please leave an explanation why you don't want to introduce such change.

ponomvrenko committed:
Added missing fetchType in `CartItem` model

ponomvrenko committed:
Cause of fetchType.Lazy in `@ManyToOne` inside `CartItem` we created SQL query that matches with `findByUserIdWithCartItems` method in `ShoppingCartRepository`.

ponomvrenko committed:
Replaced query with EntityGraph, cause of convention of my project
